hl7/containers.py

```diff
@@ -73,20 +73,260 @@ def __getslice__(self, i, j):
+        )
+
+    def create_batch(self, seq):
+        """Create a new :py:class:`hl7.Batch` compatible with this container"""
+        return self.factory.create_batch(
+            self.separators[0],
+            seq,
+            esc=self.esc,
+            separators=self.separators,
+            factory=self.factory,
+        )
+
+    def create_message(self, seq):
+        """Create a new :py:class:`hl7.Message` compatible with this container"""
+        return self.factory.create_message(
+            self.separators[0],
+            seq,
+            esc=self.esc,
+            separators=self.separators,
+            factory=self.factory,
+        )
+
+    def create_segment(self, seq):
+        """Create a new :py:class:`hl7.Segment` compatible with this container"""
+        return self.factory.create_segment(
+            self.separators[1],
+            seq,
+            esc=self.esc,
+            separators=self.separators[1:],
+            factory=self.factory,
+        )
+
+    def create_field(self, seq):
+        """Create a new :py:class:`hl7.Field` compatible with this container"""
+        return self.factory.create_field(
+            self.separators[2],
+            seq,
+            esc=self.esc,
+            separators=self.separators[2:],
+            factory=self.factory,
+        )
+
+    def create_repetition(self, seq):
+        """Create a new :py:class:`hl7.Repetition` compatible with this container"""
+        return self.factory.create_repetition(
+            self.separators[3],
+            seq,
+            esc=self.esc,
+            separators=self.separators[3:],
+            factory=self.factory,
+        )
+
+    def create_component(self, seq):
+        """Create a new :py:class:`hl7.Component` compatible with this container"""
+        return self.factory.create_component(
+            self.separators[4],
+            seq,
+            esc=self.esc,
+            separators=self.separators[4:],
+            factory=self.factory,
+        )
+
+
+class File(Container, Builder):
+    """Representation of an HL7 file from the batch protocol.
+    It contains a list of :py:class:`hl7.Batch`
+    instances. It may contain FHS/FTS :py:class:`hl7.Segment` instances.
+
+    Files may or may not be wrapped in FHS/FTS segements
+    deliniating the start/end of the batch. These are optional.
+    """
+
+    def __init__(
+        self, separator, sequence=[], esc="\\", separators="\r|~^&", factory=None
+    ):
+        super(File, self).__init__(
+            separator,
+            sequence=sequence,
+            esc=esc,
+            separators=separators,
+            factory=factory,
+        )
+        self.header = None
+        self.trailer = None
+
+    @property
+    def header(self):
+        """FHS :py:class:`hl7.Segment`"""
+        return self._batch_header_segment
+
+    @header.setter
+    def header(self, segment):
+        assert not segment or segment[0][0] == "FHS", 'header must begin with "FHS"'
```

Comment:
I know we have an `assert` statement in parser.py and a few that got added recently in streams.py, but seeing the additions here, I'm tempted to recommend that we define our own exception (or exceptions), and raise those instead of using assert statements.

1. assert is removed when the [-O (optimize)](https://docs.python.org/3/using/cmdline.html#cmdoption-o) flag is passed to Python
2. asserts are better to document thinks that will never occur, but as a library, it is feasible that someone will pass in bad data to this setter. If they wanted to catch it, they would need a `try: ... except AssertionError: ...` which feels too broad.

Rough sketch, perhaps in `hl7/exceptions.py`:
```python
class HL7Exception(Exception):
    pass

class IllegalHeader(HL7Exception): # open to a better name
    pass
```

I'd be tempted to convert the parser.py and streams.py assert statements too

Reply:
Sure. I agree, was just trying to make it feel like the rest of the library.

I'll create the exceptions.py, make sure they're exposed via __init__.py, and use them instead of asserts.

Update to the PR for this within the hour.